metric.go

```diff
@@ -113,34 +113,54 @@ type MetricDataArray []*MetricData
 
 // for ES
 type MetricDefinition struct {
-	Id         string   `json:"id"`
-	OrgId      int      `json:"org_id"`
-	Name       string   `json:"name" elastic:"type:string,index:not_analyzed"` // graphite format
-	Metric     string   `json:"metric"`                                        // kairosdb format (like graphite, but not including some tags)
-	Interval   int      `json:"interval"`                                      // minimum 10
-	Unit       string   `json:"unit"`
-	Mtype      string   `json:"mtype"`
-	Tags       []string `json:"tags" elastic:"type:string,index:not_analyzed"`
-	LastUpdate int64    `json:"lastUpdate"` // unix timestamp
-	Partition  int32    `json:"partition"`
+	Id           string   `json:"id"`
+	OrgId        int      `json:"org_id"`
+	Name         string   `json:"name" elastic:"type:string,index:not_analyzed"` // graphite format
+	Metric       string   `json:"metric"`                                        // kairosdb format (like graphite, but not including some tags)
+	Interval     int      `json:"interval"`                                      // minimum 10
+	Unit         string   `json:"unit"`
+	Mtype        string   `json:"mtype"`
+	Tags         []string `json:"tags" elastic:"type:string,index:not_analyzed"`
+	LastUpdate   int64    `json:"lastUpdate"` // unix timestamp
+	Partition    int32    `json:"partition"`
+	NameWithTags string   `json:"nameWithTags"`
 }
 
 func (m *MetricDefinition) SetId() {
 	sort.Strings(m.Tags)
 
-	buffer := bytes.NewBufferString(m.Metric)
-	buffer.WriteByte(0)
-	buffer.WriteString(m.Unit)
-	buffer.WriteByte(0)
-	buffer.WriteString(m.Mtype)
-	buffer.WriteByte(0)
-	fmt.Fprintf(buffer, "%d", m.Interval)
+	nameWithTagsBuffer := bytes.NewBufferString(m.Name)
 
-	for _, k := range m.Tags {
-		buffer.WriteByte(0)
-		buffer.WriteString(k)
+	idBuffer := bytes.NewBufferString(m.Metric)
+	idBuffer.WriteByte(0)
+	idBuffer.WriteString(m.Unit)
+	idBuffer.WriteByte(0)
+	idBuffer.WriteString(m.Mtype)
+	idBuffer.WriteByte(0)
+	fmt.Fprintf(idBuffer, "%d", m.Interval)
+
+	tagPositions := make([]int, 0, len(m.Tags)*2)
+	for _, t := range m.Tags {
+		if len(t) >= 5 && t[:5] == "name=" {
+			continue
+		}
+
+		nameWithTagsBuffer.WriteString(";")
+		tagPositions = append(tagPositions, nameWithTagsBuffer.Len())
+		nameWithTagsBuffer.WriteString(t)
+		tagPositions = append(tagPositions, nameWithTagsBuffer.Len())
+
+		idBuffer.WriteByte(0)
+		idBuffer.WriteString(t)
 	}
-	m.Id = fmt.Sprintf("%d.%x", m.OrgId, md5.Sum(buffer.Bytes()))
+
+	m.NameWithTags = fmt.Sprintf("%s", nameWithTagsBuffer)
```

Comment:
It seems odd to me that this would be a side-effect of `SetId`

Reply:
Good point. I just didn't want to change the interface.  I think in this case it would make sense to add a `NewMetricDefinition()` method, and if it isn't used then the user simply won't get the benefit of the single string backing name & tags, but it would still work as it does now.